dds/DCPS/Xcdr2ValueWriter.h

```diff
@@ -0,0 +1,187 @@
+    , encoding_(encoding)
+    , pos_(0)
+    , ser_(0)
+  {}
+
+  void begin_struct(Extensibility extensibility);
+  void end_struct();
+  void begin_struct_member(VWriterMemberParam params);
+  void end_struct_member();
+
+  void begin_union(Extensibility extensibility);
+  void end_union();
+  void begin_discriminator(VWriterMemberParam params);
+  void end_discriminator();
+  void begin_union_member(VWriterMemberParam params);
+  void end_union_member();
+
+  void begin_array(XTypes::TypeKind elem_kind);
+  void end_array();
+  void begin_sequence(XTypes::TypeKind elem_kind, ACE_CDR::ULong length);
+  void end_sequence();
+  void begin_element(size_t idx);
+  void end_element();
+
+  void write_boolean(ACE_CDR::Boolean value);
+  void write_byte(ACE_CDR::Octet value);
+#if OPENDDS_HAS_EXPLICIT_INTS
+  void write_int8(ACE_CDR::Int8 value);
+  void write_uint8(ACE_CDR::UInt8 value);
+#endif
+  void write_int16(ACE_CDR::Short value);
+  void write_uint16(ACE_CDR::UShort value);
+  void write_int32(ACE_CDR::Long value);
+  void write_uint32(ACE_CDR::ULong value);
+  void write_int64(ACE_CDR::LongLong value);
+  void write_uint64(ACE_CDR::ULongLong value);
+  void write_float32(ACE_CDR::Float value);
+  void write_float64(ACE_CDR::Double value);
+  void write_float128(ACE_CDR::LongDouble value);
+  void write_fixed(const OpenDDS::FaceTypes::Fixed& value);
+  void write_char8(ACE_CDR::Char value);
+  void write_char16(ACE_CDR::WChar value);
+  void write_string(const ACE_CDR::Char* value, size_t length);
+  void write_wstring(const ACE_CDR::WChar* value, size_t length);
+  void write_enum(const char* name, ACE_CDR::Long value, XTypes::TypeKind as_int = XTypes::TK_INT32);
+  void write_absent_value();
+
+  void write_boolean_array(const ACE_CDR::Boolean* x, size_t length);
+  void write_byte_array(const ACE_CDR::Octet* x, size_t length);
+#if OPENDDS_HAS_EXPLICIT_INTS
+  void write_int8_array(const ACE_CDR::Int8* x, size_t length);
+  void write_uint8_array(const ACE_CDR::UInt8* x, size_t length);
+#endif
+  void write_int16_array(const ACE_CDR::Short* x, size_t length);
+  void write_uint16_array(const ACE_CDR::UShort* x, size_t length);
+  void write_int32_array(const ACE_CDR::Long* x, size_t length);
+  void write_uint32_array(const ACE_CDR::ULong* x, size_t length);
+  void write_int64_array(const ACE_CDR::LongLong* x, size_t length);
+  void write_uint64_array(const ACE_CDR::ULongLong* x, size_t length);
+  void write_float32_array(const ACE_CDR::Float* x, size_t length);
+  void write_float64_array(const ACE_CDR::Double* x, size_t length);
+  void write_float128_array(const ACE_CDR::LongDouble* x, size_t length);
+  void write_char8_array(const ACE_CDR::Char* x, size_t length);
+  void write_char16_array(const ACE_CDR::WChar* x, size_t length);
+
+  size_t get_serialized_size() const;
+  const std::vector<size_t>& get_serialized_sizes() const;
+
+  // Reset the operation mode to only computing serialization size.
+  void reset()
```

Comment:
What's the expected use case for `reset()`?  Is it used in tests?

Reply:
The intended use case is to allow reusing the value writer object for serializing a different object. The usage would be like:
```
Xcdr2ValueWriter vw(xcdr2); // The default mode is computing serialized size
Sample sample;
vwrite(vw, sample); // Compute the sizes.
vw.set_serializer(ser); // Switch to serialization mode
vwrite(vw, sample); // Serialize the sample
...
Sample sample2;
vw.reset(); // Switch back to the serialized size mode
vwrite(vw, sample2);
vw.set_serializer(ser2);
vwrite(vw, sample2);
```
I think I don't have a test for it yet. I can add one.

Reply:
There are a few options: 
- no reset method, user must create a new object
- no reset method needed since it automatically changes back to the "reset" state
- provide a reset method
- others?

Reply:
I forgot to mention that right now, if the object is only used in the serialized size mode, then it doesn't need to call `reset`. This should be fine:
```
Xcdr2ValueWriter vw(xcdr2);
Sample sample;
vwrite(vw, sample);
vw.get_serialized_size(); // size of sample
Sample sample2;
vwrite(vw, sample2);
vw.get_serialized_size(); // size of sample2
```
We only need to call `reset` if the mode changes.